### PR TITLE
Websocket support for netto

### PR DESCRIPTION
**Pull request by jmcardon**

### CODE REVIEW

File: examples/src/main/scala/org/http4s/HelloWorldServer.scala [hunk 1]
```diff
@@ -2,12 +2,17 @@ package org.http4s
...
 import scala.concurrent.ExecutionContext.Implicits.global
 
 object Natto extends StreamApp[IO] {
 
   def stream(args: List[String], requestShutdown: IO[Unit]): fs2.Stream[IO, StreamApp.ExitCode] =
-    MikuBuilder[IO](new HelloWorldService[IO].appRoutes).serve
+    for {
+      sc <- Scheduler[IO](1)
+      serve <- MikuBuilder[IO](new HelloWorldService[IO](sc).appRoutes).withWebSockets(true).serve
+    } yield serve
```
Comment (emilypi) [inline examples/src/main/scala/org/http4s/HelloWorldServer.scala:16]: _nice_

File: examples/src/main/scala/org/http4s/a.sc [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+import cats.syntax.all._
+
+
+IO.async[Unit] { cb =>
+  var i = 0
+  while(i < 1000000)
+    i +=1
+  cb(Right(println("LOL1")))
+}.unsafeRunAsync(_ => ())
+println("LOL")
```
Comment (emilypi) [inline examples/src/main/scala/org/http4s/a.sc]: Prolly should remove that 😄

File: miku/src/main/scala/org/http4s/miku/MikuBuilder.scala [hunk 3]
```diff
@@ -69,17 +69,22 @@ class MikuBuilder[F[_]](
...
+    with WebSocketSupport[F] {
   implicit val e     = ec
   private val logger = getLogger
 
   protected[this] def newRequestHandler(): ChannelInboundHandler =
-    new Http4sNettyHandler[F](httpService, serviceErrorHandler) {}
+    if (enableWebsockets)
+      MikuHandler.websocket[F](httpService, serviceErrorHandler)
+    else
+      MikuHandler.default[F](httpService, serviceErrorHandler)
```
Comment (emilypi) [inline miku/src/main/scala/org/http4s/miku/MikuBuilder.scala:87]: \o/

File: miku/src/main/scala/org/http4s/miku/MikuHandler.scala [hunk 4]
```diff
@@ -47,18 +47,12 @@ abstract class Http4sNettyHandler[F[_]](service: HttpService[F], serviceErrorHan
   // in.
   private var lastResponseSent: Future[Unit] = Future.successful(())
```
Comment (emilypi) [inline miku/src/main/scala/org/http4s/miku/MikuHandler.scala:48]: Do you still need this now that you're using an `Effect`? `F.unit`?

File: miku/src/main/scala/org/http4s/miku/MikuHandler.scala [hunk 5]
```diff
@@ -168,3 +162,49 @@ abstract class Http4sNettyHandler[F[_]](service: HttpService[F], serviceErrorHan
...
+  ) extends MikuHandler[F](service, serviceErrorHandler) {
+    override def handle(channel: Channel, request: HttpRequest): F[DefaultHttpResponse] = {
+      logger.trace("Http request received by netty: " + request)
+      NettyModelConversion
+        .fromNettyRequest[F](channel, request)
+        .flatMap { request =>
+          Async.shift(ec) >> F
+            .suspend(unwrapped(request))
+            .recoverWith(serviceErrorHandler(request))
+            .flatMap(NettyModelConversion.toNettyResponseWithWebsocket[F](request, _))
```
Comment (emilypi) [inline miku/src/main/scala/org/http4s/miku/MikuHandler.scala:196]: 👍

File: miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala [hunk 6]
```diff
@@ -138,34 +137,123 @@ object NettyModelConversion {
...
+          val resp =
+            new DefaultWebSocketHttpResponse(httpVersion, HttpResponseStatus.OK, processor, factory)
+          wsContext.headers.foreach(h => resp.headers().add(h.name.toString(), h.value))
+          resp
+        }
+      }
+    } else {
+      wsContext.failureResponse.map(toNonWSResponse[F](_, httpVersion))
+    }
+
```
Comment (emilypi) [inline miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala:236]: These responses are nagging me. I think we can improve on these to some degree.

File: miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala [hunk 7]
```diff
@@ -138,34 +137,123 @@ object NettyModelConversion {
...
+  private def nettyWsToHttp4s(w: WSFrame): WebSocketFrame =
+    w match {
+      case c: TextWebSocketFrame         => Text(bytebufToArray(c.content()), c.isFinalFragment)
+      case c: BinaryWebSocketFrame       => Binary(bytebufToArray(c.content()), c.isFinalFragment)
+      case c: PingWebSocketFrame         => Ping(bytebufToArray(c.content()))
+      case c: PongWebSocketFrame         => Pong(bytebufToArray(c.content()))
+      case c: ContinuationWebSocketFrame => Continuation(bytebufToArray(c.content()), c.isFinalFragment)
+      case c: CloseWebSocketFrame        => Close(bytebufToArray(c.content()))
+    }
```
Comment (emilypi) [inline miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala:257]: These casts are nasty

File: miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala [hunk 8]
```diff
@@ -181,6 +269,13 @@ object NettyModelConversion {
...
       }
 
+  private def bytebufToArray(buf: ByteBuf): Array[Byte] = {
+    val array = new Array[Byte](buf.readableBytes())
+    buf.readBytes(array)
+    buf.release()
+    array
+  }
+
   private val CachedEmpty: DefaultHttpContent =
```
Comment (emilypi) [inline miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala:279]: I'm pretty sure you can `toArray` on a bytebuffer to turn it into an array. It should use `Arrays.copy` or `System.arrayCopy` iirc. I don't know what this `ByteBuf` is, but i would imagine you could use plain old java bytebuffers to good end here.

**Review by emilypi — COMMENTED** [review]
I'm going to take a more in depth look at the codebase when my week starts, but this looks like a great base to start with

File: examples/src/main/scala/org/http4s/a.sc (hunk 2 above)
Comment (jmcardon) [reply examples/src/main/scala/org/http4s/a.sc]: wtf. I LOL that was gone.

File: miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala (hunk 7 above)
Comment (jmcardon) [reply miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala:257]: in what way could we improve this though?

File: miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala (hunk 6 above)
Comment (jmcardon) [reply miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala:236]: This is encoded in the wsContext, for the client to have a particular response they want to send.

File: miku/src/main/scala/org/http4s/miku/MikuHandler.scala (hunk 4 above)
Comment (jmcardon) [reply miku/src/main/scala/org/http4s/miku/MikuHandler.scala:48]: yeah we do. Unfortunately, I understand _why_ they did this, but more context on it:

that `var` is essentially `threadLocal`: `lastResponseSent` runs on an `EventLoop` thread pool in netty, each thread which handles writes handed off by the main boss thread. We want writes to be executed in order, thus the entire `lastResponse = lastResponse.flatMap` nonsense makes sense from what the play people did: It's simply treating writes as a FIFO queue by using the eventloop execution context (hence we pass `trampoline`) to register the callbacks in the sequence they came in, and forking the execution of the request handler immediately.

It's sort of hard to reason about, I'm aware, but actually, the other solution is `mapAsync` from fs2 and that is a _lot_ more involved and less optimized than this quick hack

File: miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala (hunk 8 above)
Comment (jmcardon) [reply miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala:279]: `ByteBuf` is a netty construct, which you need to explicitly release. That is, imagine `free` using C.

It's not a `ByteBuffer`. If it was and I didn't have to worry about `release()`, I'd just use `Chunk.bytebuffer` but it's not :(.

**Pull request merged**